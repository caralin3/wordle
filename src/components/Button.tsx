import React from 'react';
import { Animated, StyleSheet } from 'react-native';
import { Button as RNPButton, useTheme } from 'react-native-paper';

export interface ButtonProps {
  disabled?: boolean;
  onPress: () => void;
  title: string;
}

export const Button: React.FC<ButtonProps> = ({ disabled = false, onPress, title }) => {
  const { colors } = useTheme();
  const buttonAnim = new Animated.Value(0);
  const inputRange = [0, 1];
  const outputRange = [1, 0.9];
  const scale = buttonAnim.interpolate({ inputRange, outputRange });

  const onPressIn = () => {
    Animated.spring(buttonAnim, {
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };
  const onPressOut = () => {
    Animated.spring(buttonAnim, {
      toValue: 0,
      useNativeDriver: true,
    }).start();
  };

  const handlePress = () => {
    if (!disabled) {
      onPressIn();
      setTimeout(() => {
        onPressOut();
        onPress();
      }, 50);
    }
  };

  const labelStyle = {
    color: colors.success.text,
    fontSize: 24,
    opacity: disabled ? 0.75 : 1,
  };

  return (
    <Animated.View style={[{ transform: [{ scale }] }]}>
      <RNPButton
        uppercase
        color={disabled ? colors.success.disabled : colors.success.background}
        labelStyle={labelStyle}
        onPress={handlePress}
        mode='contained'
        theme={{
          roundness: 15,
          colors: {
            disabled: colors.success.disabled,
          },
        }}
        contentStyle={styles.container}
      >
        {title}
      </RNPButton>
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 20,
    paddingVertical: 5,
  },
});
