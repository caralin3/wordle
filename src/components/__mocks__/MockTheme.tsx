import React from 'react';
import { Provider as PaperProvider } from 'react-native-paper';
import { defaultTheme } from '../../appearance';

export const MockTheme = ({ children }) => <PaperProvider theme={defaultTheme}>{children}</PaperProvider>;
