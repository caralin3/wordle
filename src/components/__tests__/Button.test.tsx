import React from 'react';
import renderer from 'react-test-renderer';
import { Button } from '../Button';
import { MockTheme } from '../__mocks__/MockTheme';

describe('Button', () => {
  test('renders correctly', () => {
    const btnLabel = 'Press';
    const mockPressFn = jest.fn();
    const tree = renderer
      .create(
        <MockTheme>
          <Button title={btnLabel} onPress={mockPressFn} />
        </MockTheme>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
