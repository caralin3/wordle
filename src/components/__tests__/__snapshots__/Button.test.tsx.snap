// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#0e8f33",
          "borderColor": "transparent",
          "borderRadius": 15,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 15,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "paddingHorizontal": 20,
                "paddingVertical": 5,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 24,
                    "opacity": 1,
                  },
                ],
              ]
            }
          >
            Press
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#0e8f33",
          "borderColor": "transparent",
          "borderRadius": 15,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 15,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "paddingHorizontal": 20,
                "paddingVertical": 5,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 24,
                    "opacity": 1,
                  },
                ],
              ]
            }
          >
            Press
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
